src/Microsoft.AspNet.Razor.Runtime/TagHelpers/TagHelperDescriptorFactory.cs

```diff
@@ -34,41 +36,170 @@ public static class TagHelperDescriptorFactory
+        /// <summary>
+        /// Internal for testing.
+        /// </summary>
+        internal static IEnumerable<string> GetCommaSeparatedValues(string text)
         {
-            var typeInfo = tagHelperType.GetTypeInfo();
-            var attributes = typeInfo.GetCustomAttributes<HtmlElementNameAttribute>(inherit: false);
+            // We don't want to remove empty entries, need to notify users of invalid values.
+            return text?.Split(',').Select(tagName => tagName.Trim()) ?? Enumerable.Empty<string>();
+        }
 
-            // If there isn't an attribute specifying the tag name derive it from the name
-            if (!attributes.Any())
-            {
-                var name = typeInfo.Name;
+        /// <summary>
+        /// Internal for testing.
+        /// </summary>
+        internal static bool ValidTargetElementAttributeNames(
+            TargetElementAttribute attribute,
+            ParserErrorSink errorSink)
+        {
+            var validTagName = ValidateName(attribute.Tag, targetingAttributes: false, errorSink: errorSink);
+            var validAttributeNames = true;
+            var attributeNames = GetCommaSeparatedValues(attribute.Attributes);
 
-                if (name.EndsWith(TagHelperNameEnding, StringComparison.OrdinalIgnoreCase))
+            foreach (var attributeName in attributeNames)
+            {
+                if (!ValidateName(attributeName, targetingAttributes: true, errorSink: errorSink))
                 {
-                    name = name.Substring(0, name.Length - TagHelperNameEnding.Length);
+                    validAttributeNames = false;
                 }
+            }
+
+            return validTagName && validAttributeNames;
+
+        }
+
+        private static bool ValidateName(
+            string name,
+            bool targetingAttributes,
+            ParserErrorSink errorSink)
+        {
+            var targetName = targetingAttributes ?
+                Resources.TagHelperDescriptorFactory_Attribute :
+                Resources.TagHelperDescriptorFactory_Tag;
+            var validName = true;
 
-                return new[] { ToHtmlCase(name) };
+            if (string.IsNullOrWhiteSpace(name))
+            {
+                errorSink.OnError(
+                    SourceLocation.Zero,
+                    Resources.FormatTargetElementAttribute_NameCannotBeNullOrWhitespace(targetName));
+
+                validName = false;
+            }
+            else
+            {
+                foreach (var character in name)
+                {
+                    if (char.IsWhiteSpace(character) ||
+                        InvalidNonWhitespaceNameCharacters.Contains(character))
+                    {
+                        errorSink.OnError(
+                            SourceLocation.Zero,
+                            Resources.FormatTargetElementAttribute_InvalidName(
+                                targetName.ToLower(),
```

Comment:
`CultureInfo.CurrentCulture` is available in Core.  `CultureInfo` might contain a bit less with that runtime but what you need is available.